### ISSUE TITLE
[BUG]:  unity csi driver v2.2.0 unable to attach pv to the nodes where the pods are scheduled

### ISSUE THREAD

**Issue by rajbaratht** [dell/csm#443]
### Bug Description

Dell unity doesn't seem attach pv to the right worker nodes. We recently replaced our worker nodes, dell unity csi driver doesn't seem to attach the pv to the right worker nodes where the pods are scheduled. I tried to restart the unity pods as well the as the application pods - nothing seems to help us. 
All the statefulset applications are down because of the this.

### Logs

Error logs from unity node and controller pods

```
k logs -l app=unity-controller -c driver | grep -i "error"
time="2022-08-25T13:40:47Z" level=info msg="/csi.v1.Controller/ControllerPublishVolume: REP 0609: rpc error: code = Aborted desc = pending"
time="2022-08-25T13:40:48Z" level=info msg="/csi.v1.Controller/ControllerPublishVolume: REP 0608: rpc error: code = Aborted desc =  runid=608 Volume has been published to a different host already."


k logs -l app=unity-node -c driver | grep -i "error"
time="2022-08-25T13:44:02Z" level=info msg="/csi.v1.Node/NodeStageVolume: REP 2041: rpc error: code = Aborted desc =  runid=2041 Volume dev2-p2-31a62a3fff has not been published to this node d2-cont-wkr1."
time="2022-08-25T13:44:10Z" level=info msg="/csi.v1.Node/NodeStageVolume: REP 6742: rpc error: code = Aborted desc =  runid=6742 Volume dev2-p2-3450e1b7b2 has not been published to this node d2-cont-wkr3."

```

### Additional Environment Information

helm values

```
helm  get values unity
USER-SUPPLIED VALUES:
allowRWOMultiPodAccess: "false"
certSecretCount: 1
controller:
  controllerCount: 2
  resizer:
    enabled: "true"
  snapshot:
    enabled: "true"
    snapNamePrefix: dev2-p2-snap
  volumeNamePrefix: dev2-p2
imagePullPolicy: Always
images:
  driverRepository: dellemc
kubeletConfigDir: /var/lib/kubelet
logLevel: debug
maxUnityVolumesPerNode: 0
node:
  dnsPolicy: ClusterFirstWithHostNet
  healthMonitor:
    enabled: false
openshift: false
podmon:
  enabled: false
  image: None
syncNodeInfoInterval: 15
tenantName: ""
version: 2.2.0
```

Describe of the pod waiting for attachment

```
[dev2-p2|data] raj@ptp-ts-raj:~$ kd po/redis-master-0
Name:           redis-master-0
Namespace:      data
Priority:       0
Node:           d2-cont-wkr3/10.40.14.250
Start Time:     Wed, 24 Aug 2022 17:51:16 -0400
Labels:         app.kubernetes.io/component=master
                app.kubernetes.io/instance=redis
                app.kubernetes.io/managed-by=Helm
                app.kubernetes.io/name=redis
                controller-revision-hash=redis-master-55bf56b75f
                helm.sh/chart=redis-16.2.0
                statefulset.kubernetes.io/pod-name=redis-master-0
Annotations:    checksum/configmap: 0cff5a11a35bea69c6e2c0fce2943df71f60a3d7f8f6e836c799e9a1d47fb2da
                checksum/health: 9e4b4ce0b6f4bdbb2060f9d986bc5a552b9a92931551a6525bc2416918d00700
                checksum/scripts: b52236bf6a1a2edf4946f38c73b79ced6323e22fd80ed761cc84fd9ec2f962b8
                checksum/secret: eb286fdaf5aa940f9648cb31c3bef23f5860f3334b08eae523d68be0cf50008c
Status:         Pending
IP:
IPs:            <none>
Controlled By:  StatefulSet/redis-master
Containers:
  redis:
    Container ID:
    Image:         docker.io/bitnami/redis:6.2.6-debian-10-r103
    Image ID:
    Port:          6379/TCP
    Host Port:     0/TCP
    Command:
      /bin/bash
    Args:
      -c
      /opt/bitnami/scripts/start-scripts/start-master.sh
    State:          Waiting
      Reason:       ContainerCreating
    Ready:          False
    Restart Count:  0
    Liveness:       exec [sh -c /health/ping_liveness_local.sh 5] delay=20s timeout=6s period=5s #success=1 #failure=5
    Readiness:      exec [sh -c /health/ping_readiness_local.sh 1] delay=20s timeout=2s period=5s #success=1 #failure=5
    Environment:
      BITNAMI_DEBUG:           false
      REDIS_REPLICATION_MODE:  master
      ALLOW_EMPTY_PASSWORD:    no
      REDIS_PASSWORD:          <set to the key 'redis-password' in secret 'redis'>  Optional: false
      REDIS_TLS_ENABLED:       no
      REDIS_PORT:              6379
    Mounts:
      /data from redis-data (rw)
      /health from health (rw)
      /opt/bitnami/redis/etc/ from redis-tmp-conf (rw)
      /opt/bitnami/redis/mounted-etc from config (rw)
      /opt/bitnami/scripts/start-scripts from start-scripts (rw)
      /tmp from tmp (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-t6ffz (ro)
Conditions:
  Type              Status
  Initialized       True
  Ready             False
  ContainersReady   False
  PodScheduled      True
Volumes:
  redis-data:
    Type:       PersistentVolumeClaim (a reference to a PersistentVolumeClaim in the same namespace)
    ClaimName:  redis-data-redis-master-0
    ReadOnly:   false
  start-scripts:
    Type:      ConfigMap (a volume populated by a ConfigMap)
    Name:      redis-scripts
    Optional:  false
  health:
    Type:      ConfigMap (a volume populated by a ConfigMap)
    Name:      redis-health
    Optional:  false
  config:
    Type:      ConfigMap (a volume populated by a ConfigMap)
    Name:      redis-configuration
    Optional:  false
  redis-tmp-conf:
    Type:       EmptyDir (a temporary directory that shares a pod's lifetime)
    Medium:
    SizeLimit:  <unset>
  tmp:
    Type:       EmptyDir (a temporary directory that shares a pod's lifetime)
    Medium:
    SizeLimit:  <unset>
  kube-api-access-t6ffz:
    Type:                    Projected (a volume that contains injected data from multiple sources)
    TokenExpirationSeconds:  3607
    ConfigMapName:           kube-root-ca.crt
    ConfigMapOptional:       <nil>
    DownwardAPI:             true
QoS Class:                   BestEffort
Node-Selectors:              <none>
Tolerations:                 node.kubernetes.io/not-ready:NoExecute op=Exists for 300s
                             node.kubernetes.io/unreachable:NoExecute op=Exists for 300s
Events:
  Type     Reason       Age                   From     Message
  ----     ------       ----                  ----     -------
  Warning  FailedMount  49m (x20 over 14h)    kubelet  Unable to attach or mount volumes: unmounted volumes=[redis-data], unattached volumes=[redis-tmp-conf tmp kube-api-access-t6ffz start-scripts health redis-data config]: timed out waiting for the condition
  Warning  FailedMount  24m (x34 over 14h)    kubelet  Unable to attach or mount volumes: unmounted volumes=[redis-data], unattached volumes=[start-scripts health redis-data config redis-tmp-conf tmp kube-api-access-t6ffz]: timed out waiting for the condition
  Warning  FailedMount  13m (x680 over 15h)   kubelet  (combined from similar events): MountVolume.MountDevice failed for volume "dev2-p2-8d3c4ec9c9" : rpc error: code = Aborted desc =  runid=6519 Volume dev2-p2-8d3c4ec9c9 has not been published to this node d2-cont-wkr3.
  Warning  FailedMount  8m23s (x33 over 15h)  kubelet  Unable to attach or mount volumes: unmounted volumes=[redis-data], unattached volumes=[config redis-tmp-conf tmp kube-api-access-t6ffz start-scripts health redis-data]: timed out waiting for the condition
  Warning  FailedMount  3m52s (x16 over 15h)  kubelet  Unable to attach or mount volumes: unmounted volumes=[redis-data], unattached volumes=[tmp kube-api-access-t6ffz start-scripts health redis-data config redis-tmp-conf]: timed out waiting for the condition
```

Describe of the pv
```
kd pv/dev2-p2-8d3c4ec9c9
Name:              dev2-p2-8d3c4ec9c9
Labels:            <none>
Annotations:       pv.kubernetes.io/provisioned-by: csi-unity.dellemc.com
Finalizers:        [kubernetes.io/pv-protection external-attacher/csi-unity-dellemc-com]
StorageClass:      unity-iscsi
Status:            Bound
Claim:             data/redis-data-redis-master-0
Reclaim Policy:    Delete
Access Modes:      RWO
VolumeMode:        Filesystem
Capacity:          8Gi
Node Affinity:
  Required Terms:
    Term 0:        csi-unity.dellemc.com/apm00202405493-iscsi in [true]
                   csi-unity.dellemc.com/apm00202405493-nfs in [true]
Message:
Source:
    Type:              CSI (a Container Storage Interface (CSI) volume source)
    Driver:            csi-unity.dellemc.com
    FSType:            ext4
    VolumeHandle:      dev2-p2-8d3c4ec9c9-iSCSI-apm00202405493-sv_452
    ReadOnly:          false
    VolumeAttributes:      arrayId=apm00202405493
                           protocol=iSCSI
                           storage.kubernetes.io/csiProvisionerIdentity=1647633013914-8081-csi-unity.dellemc.com
                           volumeId=sv_452
Events:                <none>
```

volume attachments
```
k get volumeattachments | grep "dev2-p2-8d3c4ec9c9"
csi-d7a9957a759ebfb78fba0c1ed42275a2aec5cf329194555602ea3c09dcccd4d4   csi-unity.dellemc.com   dev2-p2-8d3c4ec9c9   d2-cont-wkr7   true       2d3h
csi-eab1e11605dffb8a1278986d5da6d92e5074b62b1ee70ed809de6b768cc0be61   csi-unity.dellemc.com   dev2-p2-8d3c4ec9c9   d2-cont-wkr2   true       22h
csi-f790635e7fc8a5d6086b6e6fe9c69eefe2c48a5ee441dd0c99743aec6032c5ca   csi-unity.dellemc.com   dev2-p2-8d3c4ec9c9   d2-cont-wkr3   true       154d
```

Image showing the lun has 0 host access
![image](https://user-images.githubusercontent.com/73351998/186692403-aae4590c-c5f7-4a88-8641-5f2ff1d38d04.png)

### Steps to Reproduce

Replace a worker node with a new node, the pods eternally waiting for pv attachment

### Expected Behavior

PV should be attached/re-attached to the node where the pod is scheduled

### CSM Driver(s)

CSI Driver Unity XT v2.2.0

### Installation Type

Helm v3

### Container Orchestrator

Rancher v2.6.6 with Kubernetes Version: v1.22.10

### Operating System

Rocky Linux 8.4

**Comment by gashof** [comment]
Opened  jira https://jira.cec.lab.emc.com/browse/KRV-7118 and attached logs.

**Comment by rajendraindukuri** [comment]
Hi @rajbaratht We are looking into the issue. We will get back to you. Thanks.

**Comment by rajendraindukuri** [comment]
Hi @rajbaratht ,

We tried reproducing the scenario by deleting one of the pods of StatefulSet in our setup but we are able to see the pod scheduled on the same node.
Can you please help us with the exact steps that you followed for replacing the nodes so that we can try the same to reproduce?

Can you also please provide the attacher logs  using "kubectl logs <controller-pod-name> -n <driver-namespace> -c attacher"

Thanks
Rajendra

**Comment by rajbaratht** [comment]
@rajendraindukuri 
1. Deploy an app using statefulset
2. Make sure you have more than one worker node
3. Once when the app is up and running, cordon the node where the pod is running with the pv attachment
4. Delete the pod, the pod will be scheduled to the other worker nodes
5. Check if the unity detaches the pv from the node where the pod was running 
6. Check if the pv is attached to the new node where the pod is scheduled to run
7. Confirm the pod is in the running state

**Comment by rajbaratht** [comment]
@rajendraindukuri Yes,  I have updated the host access manually because when a pod get rescheduled to a new worker node, then the unity driver never detaches the pv and reattaches to the new node. Because of that we had to manually do it. There are currently 27 pv's in this cluster alone and we are in the beginning stages, and the number of pv's is going to increase significantly in the coming months as we ramp up the operations. If we do this manually it's gonna take us a lot of time and effort and it also affects our ability to replace/update our worker nodes.  

How do we fix this ?

**Comment by rajendraindukuri** [comment]
@rajbaratht Thanks for walking us through your env and helping us with the scenario that you are looking for. That really helped in root causing the issue. As discussed in the meeting, the issue is due to the length of the volume name not as expected by k8s.
Its a known issue in k8s and it expects the formed final volume name to be greater than or equal to 15 characters.
Looks like this got fixed in k8s 1.24 but for other versions, work around is to make sure the volume name is greater than or equal to 15 characters long.

Some of the examples are as follows:
sv-xxxx(7) +csivol(6) ----- does not work(size :13)
sv-xxxx(7) csiunity(8) ----- works(size:15)
sv-xxxxx(8) csiunity(8) ------- works(size:17)
sv-xxxxxx(9) csiunity(8) ------- works(size: 17)
sv-xxxxxx(9) csivol(6) ------- works(size : 17)

Hope this helps !!! Thanks.

**Comment by rajendraindukuri** [comment]
Closing the issue as we provided a work around for the customer on k8s version < 1.24.
